### ISSUE TITLE
Questions on MongoDB return 'No results!' after upgrade from 0.38.5 to 0.39.0 when filtering against array 

### ISSUE THREAD

**Issue by urba1n** [metabase/metabase#16299]
**Describe the bug**
After upgrading from 0.38.5 to 0.39.0 (respectively any 0.39.x), some questions for MongoDB return 'No results!'.

### Logs
Nothing in the console log.

Server logs:

```
metabase_1  | 2021-06-01 07:46:08,478 DEBUG middleware.log :: GET /api/user/current 200 4.6 ms (3 DB calls) App DB connections: 1/7 Jetty threads: 6/50 (2 idle, 0 queued) (63 total active threads) Queries in flight: 0 (0 queued)
metabase_1  | 2021-06-01 07:46:08,494 DEBUG middleware.log :: GET /api/session/properties 200 10.4 ms (2 DB calls) App DB connections: 1/7 Jetty threads: 6/50 (2 idle, 0 queued) (63 total active threads) Queries in flight: 0 (0 queued)
metabase_1  | 2021-06-01 07:46:08,514 DEBUG middleware.log :: GET /api/database 200 12.1 ms (3 DB calls) App DB connections: 0/7 Jetty threads: 5/50 (2 idle, 0 queued) (63 total active threads) Queries in flight: 0 (0 queued)
metabase_1  | 2021-06-01 07:46:08,585 DEBUG middleware.log :: GET /api/database 200 48.5 ms (4 DB calls) App DB connections: 2/7 Jetty threads: 8/50 (1 idle, 0 queued) (64 total active threads) Queries in flight: 0 (0 queued)
metabase_1  | 2021-06-01 07:46:08,587 DEBUG middleware.log :: GET /api/card/142 200 47.9 ms (5 DB calls) App DB connections: 2/7 Jetty threads: 8/50 (1 idle, 0 queued) (64 total active threads) Queries in flight: 0 (0 queued)
metabase_1  | 2021-06-01 07:46:08,596 DEBUG middleware.log :: GET /api/database 200 58.8 ms (9 DB calls) App DB connections: 1/7 Jetty threads: 8/50 (1 idle, 0 queued) (65 total active threads) Queries in flight: 0 (0 queued)
metabase_1  | 2021-06-01 07:46:08,634 DEBUG middleware.log :: GET /api/alert/question/142 200 4.0 ms (1 DB calls) App DB connections: 0/7 Jetty threads: 4/50 (3 idle, 0 queued) (65 total active threads) Queries in flight: 0 (0 queued)
metabase_1  | 2021-06-01 07:46:08,663 DEBUG middleware.log :: GET /api/table/14/query_metadata 200 18.6 ms (8 DB calls) App DB connections: 0/7 Jetty threads: 4/50 (3 idle, 0 queued) (65 total active threads) Queries in flight: 0 (0 queued)
metabase_1  | 2021-06-01 07:46:08,734 DEBUG middleware.log :: GET /api/collection/root 200 5.1 ms (2 DB calls) App DB connections: 0/7 Jetty threads: 4/50 (3 idle, 0 queued) (65 total active threads) Queries in flight: 0 (0 queued)
metabase_1  | 2021-06-01 07:46:08,895 DEBUG middleware.log :: POST /api/card/142/query 202 [ASYNC: completed] 67.1 ms (12 DB calls) App DB connections: 1/7 Jetty threads: 3/50 (3 idle, 0 queued) (66 total active threads) Queries in flight: 0 (0 queued)
```

**To Reproduce**

1. Install and setup metabase 0.38.5 and a MongoDB with a database named 'testdb' and a collection named 'testcol'.

2. Import following json in 'testcol'.
   
   ```json
   {"json_field":{"key_1": "value_jf_a", "key_2": "value_jf_z"}, "list_field":["value_lf_a", "value_lf_b"], "metas":{"group_field":"value_gf_a"}}
   {"json_field":{"key_1": "value_jf_b", "key_2": "value_jf_z"}, "list_field":["value_lf_a", "value_lf_b"], "metas":{"group_field":"value_gf_b"}}
   {"json_field":{"key_1": "value_jf_c", "key_2": "value_jf_z"}, "list_field":["value_lf_a", "value_lf_b"], "metas":{"group_field":"value_gf_c"}}
   {"json_field":{"key_1": "value_jf_a", "key_2": "value_jf_z"}, "list_field":["value_lf_c", "value_lf_d"], "metas":{"group_field":"value_gf_a"}}
   {"json_field":{"key_1": "value_jf_b", "key_2": "value_jf_z"}, "list_field":["value_lf_c", "value_lf_d"], "metas":{"group_field":"value_gf_b"}}
   {"json_field":{"key_1": "value_jf_c", "key_2": "value_jf_z"}, "list_field":["value_lf_c", "value_lf_d"], "metas":{"group_field":"value_gf_c"}}
   {"json_field":{"key_1": "value_jf_a", "key_2": "value_jf_z"}, "list_field":["value_lf_c", "value_lf_d"], "metas":{"group_field":"value_gf_a"}}
   {"json_field":{"key_1": "value_jf_b", "key_2": "value_jf_z"}, "list_field":["value_lf_c", "value_lf_d"], "metas":{"group_field":"value_gf_b"}}
   {"json_field":{"key_1": "value_jf_c", "key_2": "value_jf_z"}, "list_field":["value_lf_c", "value_lf_d"], "metas":{"group_field":"value_gf_c"}}
   ```

3. Create a 'Test question' as in screenshots

   ![screenshot_test_question_1](https://user-images.githubusercontent.com/78786/120273962-7b61bb80-c2af-11eb-9f0a-4b51b47ae01e.png)

   and

   ![screenshot_test_question_2](https://user-images.githubusercontent.com/78786/120273971-7e5cac00-c2af-11eb-8835-9f06a0e1befa.png)

   The native query for this question is :

   ```json
   [
     {
       "$project": {
         "json_field___key_1": "$json_field.key_1",
         "list_field": "$list_field",
         "metas___group_field": "$metas.group_field"
       }
     },
     {
       "$match": {
         "$and": [
           {
             "$or": [
               {
                 "json_field___key_1": {
                   "$eq": "value_jf_a"
                 }
               },
               {
                 "json_field___key_1": {
                   "$eq": "value_jf_b"
                 }
               }
             ]
           },
           {
             "list_field": {
               "$eq": "value_lf_a"
             }
           }
         ]
       }
     },
     {
       "$project": {
         "_id": "$_id",
         "___group": {
           "metas___group_field": "$metas___group_field"
         }
       }
     },
     {
       "$group": {
         "_id": "$___group",
         "count": {
           "$sum": 1
         }
       }
     },
     {
       "$sort": {
         "_id": 1
       }
     },
     {
       "$project": {
         "_id": false,
         "metas___group_field": "$_id.metas___group_field",
         "count": true
       }
     },
     {
       "$sort": {
         "metas___group_field": 1
       }
     }
   ]
   ```
   
4. The visualize result is

   ![screenshot_0_38_5](https://user-images.githubusercontent.com/78786/120273978-80bf0600-c2af-11eb-802c-06569e25af97.png)

5. Upgrade metabase to 0.39.0 (or any 0.39.x)

6. The native query for 'Test question' is
 
   ```json
   [
     {
       "$match": {
         "$and": [
           {
             "$or": [
               {
                 "$expr": {
                   "$eq": [
                     "$json_field.key_1",
                     "value_jf_a"
                   ]
                 }
               },
               {
                 "$expr": {
                   "$eq": [
                     "$json_field.key_1",
                     "value_jf_b"
                   ]
                 }
               }
             ]
           },
           {
             "$expr": {
               "$eq": [
                 "$list_field",
                 "value_lf_a"
               ]
             }
           }
         ]
       }
     },
     {
       "$group": {
         "_id": {
           "metas": {
             "group_field": "$metas.group_field"
           }
         },
         "count": {
           "$sum": 1
         }
       }
     },
     {
       "$sort": {
         "_id": 1
       }
     },
     {
       "$project": {
         "_id": false,
         "metas.group_field": "$_id.metas.group_field",
         "count": true
       }
     },
     {
       "$sort": {
         "metas.group_field": 1
       }
     }
   ]
   ```
    
   and the visualize result is screenshot_0_39_0

   ![screenshot_0_39_0](https://user-images.githubusercontent.com/78786/120273991-84528d00-c2af-11eb-8500-f7b2703352d9.png)

### Expected behavior

Have the same visualize result in 0.39.x as in 0.38.5

### Information about your Metabase Installation:

- Your browser and the version: Chrome 91.0.4472.77
- Your operating system: OS X 11.4
- Your databases: MongoDB
- Metabase version: 0.38.5 and 0.39.0 (still reproducible in every 0.39.x)
- Metabase hosting environment: Docker
- Metabase internal database: MariaDB

```json
{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.10+9",
    "java.vendor": "AdoptOpenJDK",
    "java.vendor.url": "https://adoptopenjdk.net/",
    "java.version": "11.0.10",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.10+9",
    "os.name": "Linux",
    "os.version": "5.10.25-linuxkit",
    "user.language": "en",
    "user.timezone": "Europe/Zurich"
  },
  "metabase-info": {
    "databases": [
      "h2",
      "mysql",
      "mongo"
    ],
    "hosting-env": "unknown",
    "application-database": "mysql",
    "application-database-details": {
      "database": {
        "name": "MariaDB",
        "version": "10.5.9-MariaDB-1:10.5.9+maria~focal"
      },
      "jdbc-driver": {
        "name": "MariaDB Connector/J",
        "version": "2.6.2"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2021-04-19",
      "tag": "v0.39.0",
      "branch": "release-x.39.x",
      "hash": "5e40366"
    },
    "settings": {
      "report-timezone": "Europe/Berlin"
    }
  }
}
```

### Severity

Blocking an upgrade

### Additional context

Thanks for your support

**Comment by flamber** [comment]
@urba1n Excellent issue with simple reproduction steps :+1:

**Comment by camsaul** [comment]
formatted/readable:

### before:

```json
[
  {
    "$project": {
      "json_field___key_1": "$json_field.key_1",
      "list_field": "$list_field",
      "metas___group_field": "$metas.group_field"
    }
  },
  {
    "$match": {
      "$and": [
        {
          "$or": [
            {
              "json_field___key_1": {
                "$eq": "value_jf_a"
              }
            },
            {
              "json_field___key_1": {
                "$eq": "value_jf_b"
              }
            }
          ]
        },
        {
          "list_field": {
            "$eq": "value_lf_a"
          }
        }
      ]
    }
  },
  {
    "$project": {
      "_id": "$_id",
      "___group": {
        "metas___group_field": "$metas___group_field"
      }
    }
  },
  {
    "$group": {
      "_id": "$___group",
      "count": {
        "$sum": 1
      }
    }
  },
  {
    "$sort": {
      "_id": 1
    }
  },
  {
    "$project": {
      "_id": false,
      "metas___group_field": "$_id.metas___group_field",
      "count": true
    }
  },
  {
    "$sort": {
      "metas___group_field": 1
    }
  }
]
```

### after:

```json
[
  {
    "$match": {
      "$and": [
        {
          "$or": [
            {
              "$expr": {
                "$eq": [
                  "$json_field.key_1",
                  "value_jf_a"
                ]
              }
            },
            {
              "$expr": {
                "$eq": [
                  "$json_field.key_1",
                  "value_jf_b"
                ]
              }
            }
          ]
        },
        {
          "$expr": {
            "$eq": [
              "$list_field",
              "value_lf_a"
            ]
          }
        }
      ]
    }
  },
  {
    "$group": {
      "_id": {
        "metas": {
          "group_field": "$metas.group_field"
        }
      },
      "count": {
        "$sum": 1
      }
    }
  },
  {
    "$sort": {
      "_id": 1
    }
  },
  {
    "$project": {
      "_id": false,
      "metas.group_field": "$_id.metas.group_field",
      "count": true
    }
  },
  {
    "$sort": {
      "metas.group_field": 1
    }
  }
]
```

**Comment by camsaul** [comment]
I think we just need to add the nested filtered columns to the `:fields` clause so they're included in the results -- this happened with the old code before 39 optimizations (unintentionally) -- I think it makes sense for nested columns that get filtered to be included in `:fields` if `:fields` isn't explicitly specified

**Comment by camsaul** [comment]
Ok, so the actual issue here is that the old query has 

```clj
{:list_field {:$eq "value_lf_a"}}
```

for the `:=` filters against the list fields, while the new query has

```clj
{:$expr {:$eq ["$list_field" "value_lf_a"]}}
```

the old one works, but the new one does not.

**Comment by urba1n** [comment]
Thank you very much @camsaul, I have tested branch fix-16299 with my real data and everything is working as expected.

**Comment by camsaul** [comment]
Fixed by #16480
